### PR TITLE
JIT cthelpers to query annotations

### PR DESCRIPTION
**Pull request by tajila**
JIT cthelpers to query annotations

Details are in the issues below.

Fixes: #11990

Most of the work was done by @sharon-wang. Creating this PR on her
behalf.

Signed-off-by: Tobi Ajila <atobia@ca.ibm.com>

### CODE REVIEW

**Comment by tajila** [comment]
@babsingh please review these changes

**Comment by babsingh** [comment]
@fengxue-IS refer at `methodContainsRuntimeAnnotation` which is on `line 230` of `annhelp.c`.

**Review by babsingh — CHANGES_REQUESTED** [review]
This is my first pass, which covers formatting issues. I will go through the code logic next.

File: runtime/util/annhelp.c
Comment (babsingh) [inline runtime/util/annhelp.c:305]:
```suggestion
	U_8 *dataEnd = (U_8 *)(data + length);
```

File: runtime/util/annhelp.c
Comment (babsingh) [inline runtime/util/annhelp.c:297]: `length` is missing in the function comment

File: runtime/util/annhelp.c
Comment (babsingh) [inline runtime/util/annhelp.c:340]:
```suggestion
			/* Disregard value and only iterate through the members. */
```

File: runtime/util/annhelp.c
Comment (babsingh) [inline runtime/util/annhelp.c:312]: use complete word `result` or `found`
```suggestion
	BOOLEAN found = FALSE;
```

File: test/functional/Java8andUp/src_110_up/org/openj9/test/annotation/ContainsRuntimeAnnotationTest.java
Comment (babsingh) [inline test/functional/Java8andUp/src_110_up/org/openj9/test/annotation/ContainsRuntimeAnnotationTest.java:93]: The test contains a lot of empty lines, which can be removed. Example:

```suggestion
		cpIndex = getMemberCPIndex("myField2", "I", true); //$NON-NLS-1$ //$NON-NLS-2$
		Assert.assertTrue(-1 != cpIndex, "could fine myField2");
```

**Review by babsingh — CHANGES_REQUESTED** [review]
this is my final pass.

++ @llxia for reviewing test changes.

File: test/functional/Java8andUp/src_110_up/org/openj9/test/annotation/ContainsRuntimeAnnotationTest.java
Comment (llxia) [reply test/functional/Java8andUp/src_110_up/org/openj9/test/annotation/ContainsRuntimeAnnotationTest.java:44]: yes, NON-NLS comments are not needed for tests.

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by gita-omr** [comment]
Could we try to merge this PR this week since JIT needs some time to start using these two helpers and we would like to get it into 0.27?

File: runtime/jit_vm/cthelpers.cpp
Comment (tajila) [reply runtime/jit_vm/cthelpers.cpp:199]: windows complains about the conversion from `BOOLEAN` to `bool`

File: runtime/jit_vm/cthelpers.cpp
Comment (tajila) [reply runtime/jit_vm/cthelpers.cpp:197]: not going to duplicate the function decsriptor here, makes refactoring more difficult in the future.

**Comment by tajila** [comment]
@babsingh Ive addressed the review comments, please take another look

**Review by babsingh — APPROVED** [review]
Some of the feedback in `annhelp.c` and `ContainsRuntimeAnnotationTest.java` was not addressed. This feedback mostly focuses on formatting. Due to the urgent need for this PR, addressing the formatting feedback is not critical. I am tentatively approving this PR. Jenkins test builds still need to be run. Delegating to committers: @pshipton (senior reviewer) and @llxia (test changes).

**Comment by pshipton** [comment]
jenkins test sanity win,zlinux jdk8,jdk11

File: runtime/util/annhelp.c
Comment (gacholio) [reply runtime/util/annhelp.c:306]: Personally, I don't add that space.

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/6064/

**Comment by pshipton** [comment]
jenkins compile win32 jdk8

**Comment by gacholio** [comment]
Is this ready for review? There are a pile of unresolved comments. I do not agree that we should commit it broken and fix it later. 

This is new feature work, so I don't see how it can be so urgent.

**Comment by pshipton** [comment]
@gacholio the outstanding comments are about formatting and comments, the code is ready for review.

The urgency comes from the desire to get the feature work into the 0.27 release, for which the code split is the weekend of June 6.

**Comment by gacholio** [comment]
General comment - fields and methods do not have CP indices, field and method refs do.

**Comment by pshipton** [comment]
I've resolved all the comments that were actually resolved.

**Comment by pshipton** [comment]
Note the test failures. I see lots of this
```
Cannot find class in classpath: org.openj9.test.annotation.ContainsRuntimeAnnotationTest
Exception in thread "main" java.lang.NullPointerException
	at org.testng.TestNG.getStatus(TestNG.java:211)
	at org.testng.TestNG.main(TestNG.java:1324)
```

File: runtime/util/annhelp.c
Comment (pshipton) [inline runtime/util/annhelp.c:242]: Babneet called this out but it didn't get fixed. `romClass` can be used rather than `clazz->romClass`, since `romClass` is set on the previous line.

**Comment by gacholio** [comment]
Is this intended only to work with local fields (fields declared in the same class that they are used in)? That's how the code reads to me - this doesn't sound like it's what we want.

File: runtime/util/annhelp.c
Comment (gacholio) [reply runtime/util/annhelp.c:306]: Actually, the correct change is to remove the cast entirely. The types are already correct.

**Comment by tajila** [comment]
> Is this intended only to work with local fields (fields declared in the same class that they are used in)? That's how the code reads to me - this doesn't sound like it's what we want.

@gita-omr ^^

**Comment by gacholio** [comment]
Let me be explicit - if class A references stable field B.f, it will not be detected by this code. This is true for fields inherited from superclasses and fields in unrelated classes.

**Comment by tajila** [comment]
I see `getMethodFromMethodRef` and `getFieldShapeFromFieldRef` will fail to find it

**Comment by tajila** [comment]
I think I can address that by searching in the class specified by `romFieldRef->classRefCPIndex` and `romMethodRef->classRefCPIndex`

**Comment by tajila** [comment]
> Note the test failures. I see lots of this
```
Cannot find class in classpath: org.openj9.test.annotation.ContainsRuntimeAnnotationTest
Exception in thread "main" java.lang.NullPointerException
	at org.testng.TestNG.getStatus(TestNG.java:211)
	at org.testng.TestNG.main(TestNG.java:1324)
```

I've fixed the test issues

**Comment by tajila** [comment]
I'll post when changes are ready for another look

**Comment by gacholio** [comment]
The correct thing to do is probably to resolve the ref into private memory and look there.

**Comment by gita-omr** [comment]
> > Is this intended only to work with local fields (fields declared in the same class that they are used in)? That's how the code reads to me - this doesn't sound like it's what we want.
> 
> @gita-omr ^^

I have not read about such limitation anywhere. I always thought that Stable is the same as final except that it can have two values: null and some final value.

**Comment by tajila** [comment]
@gita-omr Are we guaranteed that these helpers will only be used on resolved fields/methods?

**Comment by gacholio** [comment]
You may also need to know the type of the thing you're looking for (static/instance field, virtual/special/static/interface method) in order to decode the resolved information.

**Comment by gacholio** [comment]
In fact, some of the cases make no sense (without a receiver, you can't find the method for a virtual call). I'm not convinced these APIs make any sense - perhaps the JIT needs to determine the field or method in question, then query that rather than trying to go from the ref.

**Comment by tajila** [comment]
> In fact, some of the cases make no sense (without a receiver, you can't find the method for a virtual call). I'm not convinced these APIs make any sense - perhaps the JIT needs to determine the field or method in question, then query that rather than trying to go from the ref.

@gacholio I did a search of the code base. By convention these annotations are only used on final and static methods

**Comment by gacholio** [comment]
If this only applies to static methods and fields then the APIs make more sense. I don't see any response to the question about resolution.

**Comment by tajila** [comment]
> If this only applies to static methods and fields then the APIs make more sense. I don't see any response to the question about resolution.

@gacholio 
There is no easy way for the JIT to identify whether a callsite is monomorphic or not (at least not when they will be calling this API) so we will put the method annotations on hold for the time being and try to determine another way to approach that.

For the field annotations we can handle statics as you said. What is the issue with instance fields? A fieldref can only refer to a single field.

Even in this case:

```
class A {
		int x;
	}

	class B extends A {
		int x;
        }
```

A method in `class B` must use `super.x` to refer to `A.x`

**Comment by gacholio** [comment]
No problem with instance fields - you will have to do the resolve to find it, regardless (there's no other way to find the field declaration). The code you had is essentially the resolve code without handling superclasses and any other cases, so it makes no sense to duplicate it here.

**Comment by gita-omr** [comment]
> @gita-omr Are we guaranteed that these helpers will only be used on resolved fields/methods?

Yes, I think we can check that before calling the helper.

**Comment by tajila** [comment]
> The code you had is essentially the resolve code without handling superclasses and any other cases, so it makes no sense to duplicate it here.

Do you mean use `resolveInstanceFieldRefInto` and `resolveStaticFieldRefInto` here instead of `findFieldAndCheckVisibility` ?

The resolve code (`resolveInstanceFieldRefInto` and `resolveStaticFieldRefInto`) resolves the `romFieldRef->classRefCPIndex` classref then uses that (the declaring class) to lookup the field in `findFieldAndCheckVisibility` via resolvefield helpers. The rest of the resolution code performs access, init and visibility checks which we dont need. Essentially, after we have the fieldshape we are done. So calling `findFieldAndCheckVisibility` seems like the simplest approach here. This handles cases where the field is defined in the super class as well.

Or where you referring to the `instanceFieldOffsetWithSourceClass` and `instanceFieldOffset` helpers in resolveField?

**Comment by gacholio** [comment]
I think I was referring to older versions of the code (possibly confused with the method code). `findFieldAndCheckVisibility` should be good enough (since it's resolved we don't need to check visibility).

**Comment by tajila** [comment]
@gacholio changes are ready for another look

**Review by llxia — APPROVED** [review]
The test change looks good to me.

File: runtime/oti/j9nonbuilder.h
Comment (gacholio) [inline runtime/oti/j9nonbuilder.h:4783]: This does not match the function name, which every other entry does.

File: runtime/vm/intfunc.c
Comment (gacholio) [inline runtime/vm/intfunc.c:401]: All entries should end with a comma.

File: runtime/vm/resolvefield.cpp
Comment (gacholio) [inline runtime/vm/resolvefield.cpp:80]: This seems unnecessary - why not just make the real thing non-static?

File: runtime/util/annhelp.c
Comment (gacholio) [inline runtime/util/annhelp.c:184]: These lines appear way overindented in the web interface.

**Comment by gacholio** [comment]
Looks good modulo my nitpicks.

File: runtime/util/annhelp.c
Comment (gacholio) [inline runtime/util/annhelp.c:156]: The doc (and probably function name) should indicate that the field ref must be resolved. You may also want to add some asserts to that end.

File: runtime/util/annhelp.c
Comment (gacholio) [inline runtime/util/annhelp.c:177]: This is not the declaring class of the field, it's the class where we start the search. `fieldClass` might be better.

Comment (gacholio) [reply runtime/util/annhelp.c:177]: Or 'declaredClass` perhaps.

File: runtime/vm/resolvefield.cpp
Comment (tajila) [reply runtime/vm/resolvefield.cpp:80]: There is another definition of findField in cthelper

Comment (gacholio) [reply runtime/vm/resolvefield.cpp:80]: It's static and in another module, so I doubt it will cause any issue. If anything, I'd be tempted to rename that one rather than adding the wrapper here.

**Comment by tajila** [comment]
@gacholio I've updated the PR

**Review by gacholio — APPROVED** [review]

**Comment by gacholio** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by tajila** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by tajila** [comment]
jenkins test sanity,extended plinux jdk11

**Comment by tajila** [comment]
Forgot to fix up the java8 tests, should be fine now.

**Comment by gacholio** [comment]
https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_s390x_linux_Personal_testList_0/2/
```
17:39:04  FAILED test targets:
17:39:04  	RuntimeAnnotationTests_0
```

**Comment by tajila** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by tajila** [comment]
jenkins test sanity,extended plinux jdk11

**Comment by tajila** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by tajila** [comment]
jenkins test sanity,extended plinux jdk11

**Comment by tajila** [comment]
jenkins test sanity,extended plinux jdk11

**Comment by gacholio** [comment]
`RuntimeAnnotationTests_0` continues to fail.

**Comment by tajila** [comment]
The load lib path is wrong `export LD_LIBRARY_PATH="/home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_ppc64le_linux_Personal_testList_0/openjdkbinary/openjdk-test-image/openj9:/home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_ppc64le_linux_Personal_testList_0/openjdkbinary/j2sdk-image/bin/../lib/default:/home/jenkins/workspace`

not sure how that happened

**Comment by tajila** [comment]
jenkins test sanity,extended plinux jdk11

**Comment by tajila** [comment]
jenkins test sanity,extended plinux jdk11

**Comment by tajila** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by gacholio** [comment]
Is this ready for re-test?

**Comment by tajila** [comment]
Im testing internally, turns out I didnt update the right cmake files. Ill respin the PR builds later today

**Comment by tajila** [comment]
still having trouble with the java lib path on cmake builds. As to not delay things for the JIt team any further I have removed the tests. I will work on adding, the mean time the JIT team can start to make use of the core implementation

**Comment by tajila** [comment]
jenkins test sanity,extended plinux jdk11

**Comment by tajila** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by tajila** [comment]
jenkins test sanity,extended plinux jdk11

**Comment by tajila** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by gacholio** [comment]
Builds have all been aborted.

**Comment by gacholio** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by gacholio** [comment]
jenkins compile win jdk16

**Comment by gacholio** [comment]
I can't determine what test failed.

**Comment by llxia** [comment]
cmdLineTester_pltest_0 failed https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_s390x_linux_Personal/11/tapResults/

```
...
>> Success condition was not found: [Output match: ALL TESTS COMPLETED AND PASSED]
>> Failure condition was not found: [Output match: ASSERTION FAILED]


---TEST RESULTS---
Number of PASSED tests: 0 out of 1
Number of FAILED tests: 1 out of 1

---SUMMARY OF FAILED TESTS---
PLTEST
-----------------------------


cmdLineTester_pltest_0_FAILED
```

**Comment by pshipton** [comment]
The zlinux pltest problem is https://github.com/eclipse-openj9/openj9/issues/12719

**Comment by gita-omr** [comment]
@tajila I tried it and it works well. I have the following questions:

1) should we put the jitIsFieldStable prototype into some header?
2) Is isStatic argument needed?

**Comment by gita-omr** [comment]
Got a crash here:

```
#13 <signal handler called>
#14 0x000073b703d49780 in fieldContainsRuntimeAnnotation (currentThread=0x4ea00, clazz=0x29eb00, cpIndex=115,
    annotationName=0x73b703e51912 <ojdk_stable>)
    at /home/koblents/jvm/J2SE_OPENJ9/openj9-openjdk-jdk16/openj9/runtime/util/annhelp.c:178
#15 0x000073b703d45c60 in jitIsFieldStable (currentThread=<optimized out>, clazz=<optimized out>, cpIndex=<optimized out>,
    isStatic=<optimized out>) at /home/koblents/jvm/J2SE_OPENJ9/openj9-openjdk-jdk16/openj9/runtime/jit_vm/cthelpers.cpp:201
--Type <RET> for more, q to quit, c to continue without paging--q
Quit
(gdb) p romFieldRef
$10 = (J9ROMFieldRef *) 0x73b69d8b3568
(gdb) p romFieldRef->classRefCPIndex
$11 = 1073872912
(gdb) p clazz
$12 = (J9Class *) 0x29eb00
(gdb)

```

**Comment by gita-omr** [comment]
Maybe I am somehow providing wrong class, looking into it...

**Comment by gita-omr** [comment]
Yes, I was passing declaring class instead of referencing. Works now!

**Comment by gacholio** [comment]
>     1. should we put the jitIsFieldStable prototype into some header?

Yes, we should be providing this prototype - the other from this file are in j9protos.h

>     2. Is isStatic argument needed?

It appears unused. We're not resolving the field or using the resolved information, so I think it can be removed.

Suggest opening a new PR to change and export the function prototype before committing any JIT code against the new function.

**Comment by gita-omr** [comment]
> 
> 
> > @gita-omr Are we guaranteed that these helpers will only be used on resolved fields/methods?
> 
> Yes, I think we can check that before calling the helper.

Actually, we change the implementation to accommodate more cases and now we don't check if the field is resolved or not. Do you think it's ok? @tajila

**Comment by gita-omr** [comment]
> We're not resolving the field or using the resolved information, so I think it can be removed.

Just saw this, so it answers my question.

**Pull request merged**
